### PR TITLE
[RWRoute] Yet more cleanup

### PR DESCRIPTION
**Pull request by eddieh-xlnx**
(repeat of #1072)

- `RouteNode` reduced from 80 bytes to 64 bytes
- `RouteNodeGraph.{preservedMap,nodesMap}` (no longer `ConcurrentHashMap`-s) values' to be arrays to be sized no larger than number of base wires of its tile
- SLR crossing optimizations and distance computation fixes
- Mark dead-end nodes (i.e. those without children) with the new `RouteNodeType.INACCESSIBLE` to prevent unnecessary exploration
- Various other optimizations

Results for just the RWRoute wall time and the number of nodes it popped, for the FPGA24 Routing Contest benchmarks. 
```
Before (master @ 9cd25a6)
+--------------------------------+------+---------------+--------------+
| Benchmark                      | Pass | RWRoute (sec) | Nodes Popped |
+--------------------------------+------+---------------+--------------+
| logicnets_jscl_rwroute         | True |         39.73 |     55611142 |
| boom_med_pb_rwroute            | True |        244.56 |    250431531 |
| vtr_mcml_rwroute               | True |        238.97 |    319777647 |
| rosetta_fd_rwroute             | True |        143.65 |    189200371 |
| corundum_25g_rwroute           | True |        225.74 |    267832074 |
| finn_radioml_rwroute           | True |         93.82 |    115969076 |
| vtr_lu64peeng_rwroute          | True |        191.54 |    224923115 |
| corescore_500_rwroute          | True |        123.14 |    122451976 |
| corescore_500_pb_rwroute       | True |        246.03 |    244333747 |
| mlcad_d181_lefttwo3rds_rwroute | True |       2021.97 |   1686723818 |
| koios_dla_like_large_rwroute   | True |        280.14 |    299644675 |
| boom_soc_rwroute               | True |       1241.37 |   1449104918 |
| ispd16_example2_rwroute        | True |        472.75 |    523581515 |
+--------------------------------+------+---------------+--------------+

After @ 425b0b7
+--------------------------------+------+---------------+--------------+
| Benchmark                      | Pass | RWRoute (sec) | Nodes Popped |
+--------------------------------+------+---------------+--------------+
| logicnets_jscl_rwroute         | True |         38.81 |     54493698 |
| boom_med_pb_rwroute            | True |        233.96 |    272740956 |
| vtr_mcml_rwroute               | True |        254.06 |    339041410 |
| rosetta_fd_rwroute             | True |        149.72 |    191115169 |
| corundum_25g_rwroute           | True |        251.56 |    278838734 |
| finn_radioml_rwroute           | True |         90.17 |    115746783 |
| vtr_lu64peeng_rwroute          | True |        182.57 |    215449340 |
| corescore_500_rwroute          | True |        118.19 |    122093928 |
| corescore_500_pb_rwroute       | True |        240.86 |    239000597 |
| mlcad_d181_lefttwo3rds_rwroute | True |       1926.27 |   1730785359 |
| koios_dla_like_large_rwroute   | True |        273.57 |    299273289 |
| boom_soc_rwroute               | True |       1235.45 |   1503721309 |
| ispd16_example2_rwroute        | True |        452.05 |    514270796 |
+--------------------------------+------+---------------+--------------+
```
2.1% reduction in total wall time, despite 2.2% increase in number of nodes popped.

### CODE REVIEW

File: src/com/xilinx/rapidwright/rwroute/NetWrapper.java [hunk 1]
```diff
@@ -180,4 +180,34 @@ public SitePinInst getOrCreateAlternateSource(RouteNodeGraph routingGraph) {
     public RouteNode getAltSourceRnode() {
         return altSourceRnode;
     }
+
+    public boolean hasMultipleDrivers(int sequence) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/NetWrapper.java:184]: Is this called anywhere?

File: src/com/xilinx/rapidwright/rwroute/RouteNode.java [hunk 2]
```diff
@@ -52,7 +52,7 @@ public class RouteNode extends Node implements Comparable<RouteNode> {
     public static final int initialHistoricalCongestionCost = 1;
 
     /** The type of a rnode*/
-    private RouteNodeType type;
+    private byte type;
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNode.java:55]: Nice

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java [hunk 3]
```diff
@@ -114,20 +116,28 @@ public class RouteNodeGraph {
...
     protected final boolean isVersal;
 
-    public RouteNodeGraph(Design design, RWRouteConfig config) {
-        this(design, config, new HashMap<>());
+    protected final Map<TileTypeEnum, Integer> baseWireCounts;
+
+    protected final static int MAX_OCCUPANCY = 256;
+    protected final float[] presentCongestionCosts;
+
+    protected static int getTileCount(Design design) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java:124]: Should we add this to the `Device` object?

**Review by clavin-xlnx — APPROVED** [review]
Thanks for the improvements!

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java (hunk 3 above)
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java:124]: Yes, I think that's a good idea. Specifically, I think we should update the documentation to state that `Tile.getUniqueAddress()` should not exceed a new method that is `Tile.getMaxUniqueAddress()`.

Let's not hold up this PR until the next release though.

**Pull request merged**